LaTeX/Inhalt/04_Inhalt/unit_tests.tex

```diff
@@ -1 +1,25 @@
+
+\subsection{Repeatable}
+Unsere Tests sind wiederholbar. Sie können also immer wieder ausgeführt werden und liefern immer das gleiche Ergebnis, sofern keine Änderungen am Code vorgenommen wurden. Dies liegt daran, dass die Tests nicht von äußeren Umständen abhängen. Weitere Einblicke wie wir die Tests von äußeren Abhängigkeiten lösen, werden in der Sektion ''Mocks'' gegeben.
+
+\subsection{Independent}
+Jeder unserer Tests testet eine dedizierte Funktionalität wie bspw. das Lesen eines Eintrags aus der Datenbank oder das Ver- und Entschlüsseln von Passwörtern. Hierbei wird das Ver- und Entschlüsseln, wie man in \href{https://github.com/moorts/Morik/blob/main/src/tests/plugins/encryption/crypto_tests.cpp}{\textit{crypto\_tests}} sieht, als eine Funktionalität aufgefasst, da man ein Passwort nur entschlüsseln kann wenn es zuvor verschlüsselt wurde und das Verschlüsseln eines Passworts ohne die Intention es wieder zu entschlüsseln sinnfrei ist. Da jeder Test genau eine Funktionalität testet, gibt es keine Abhängigkeiten zwischen den Tests, weshalb auch die Ausführungsreihenfolge irrelevant ist.
+
+\subsection{Professional}
+Der Testcode von Morik ist professionell geschrieben. Dies bedeutet, dass er leicht verständlich ist aufgrund von sprechenden Bezeichnern und dem Befolgen der AAA-Normalform. Außerdem werden keine unrelevanten Aspekte der Applikation getestet, sondern wir fokussieren uns, wie bereits bei ''Thorough'' genannt, auf das Testen der wichtigsten Funktionalitäten.
+
+\section{Code Coverage}
+Wie zuvor bereits genannt, ermöglicht die Code Coverage eine Einschätzung wie ''gründlich'' die Tests tatsächlich sind, indem gemessen wird wie viel Code beim Testen durchlaufen wird im Vergleich zu wie viel Code insgesamt in der kompletten Applikation durchlaufen werden kann. Hierbei werden zwei Metriken besonders häufig verwendet. Zum einen gibt es die Line Coverage, die die ausgeführten Codezeilen zählt, während es zum anderen die Branch Coverage gibt. Diese zählt die ausgeführten Verzweigungsmöglichkeiten. Dabei ist letztere Metrik aussagekräftiger für die tatsächliche ''Gründlichkeit'' der Tests, während die Line Coverage allerdings leichter zu messen ist.\\
```

Comment:
Is this accurate? I guess when a program has many early returns, branch coverage will differ greatly, but as can be seen in our case, the metrics are extremely similar. Also, in certain programs, branch coverage could even be misleading if most branches are barely called.

My first instinct is also to say that branch coverage is more accurate, but maybe we should cite literature or something to support that claim.

Reply:
Lindner said that in his lecture, which is why I didn't back that claim by any source. Also if a program has many early returns those would probably be in a branch themselves which means that this branch as well as the other branch without the early return would need execution for coverage to reach 100%, whereas line coverage can go very high without ever executing a specific branch if that branch has only few lines. Therefore branch coverage is more accurate.